src/runtime/xrt.cpp

```diff
@@ -0,0 +1,628 @@
+                                 halide_buffer_t *buf) {
+    int ret;
+    debug(user_context)
+        << "XRT: halide_xrt_copy_to_host (user_context: " << user_context
+        << ", buf: " << buf << ")\n";
+
+    XrtBufferHandle *handle = (XrtBufferHandle *)buf->device;
+
+    XRTContext context(user_context);
+    if (context.error_code)
+        return context.error_code;
+
+    debug(user_context)
+        << "buf->size_in_bytes(): " << (uint64_t)buf->size_in_bytes()
+        << ", handle->size: " << (uint64_t)handle->size << "\n";
+
+    ret = xrtBOSync(handle->handle, XCL_BO_SYNC_BO_FROM_DEVICE, buf->size_in_bytes(), 0);
+    if (ret) {
+        error(user_context) << "XRT: halide_xrt_copy_to_host: xrtBOSync failed: "
+                            << ret << "\n";
+        return halide_error_code_generic_error;
+    }
+
+    ret = xrtBORead(handle->handle, buf->host, buf->size_in_bytes(), 0);
+    if (ret) {
+        error(user_context) << "XRT: halide_xrt_copy_to_host: xrtBORead failed: "
+                            << ret << "\n";
+        return halide_error_code_generic_error;
+    }
+
+    return halide_error_code_success;
+}
+
+WEAK int halide_xrt_device_crop(void *user_context,
+                                const struct halide_buffer_t *src,
+                                struct halide_buffer_t *dst) {
+    return halide_error_code_generic_error;
+}
+
+WEAK int halide_xrt_device_slice(void *user_context,
+                                 const struct halide_buffer_t *src,
+                                 int slice_dim,
+                                 int slice_pos,
+                                 struct halide_buffer_t *dst) {
+    return halide_error_code_generic_error;
+}
+
+WEAK int halide_xrt_device_release_crop(void *user_context,
+                                        struct halide_buffer_t *buf) {
+    return halide_error_code_generic_error;
+}
+
+WEAK int halide_xrt_wrap_native(void *user_context, struct halide_buffer_t *buf, uint64_t mem) {
+    halide_debug_assert(user_context, false && "unimplemented");
+    return halide_error_code_unimplemented;
+}
+
+WEAK int halide_xrt_detach_native(void *user_context, halide_buffer_t *buf) {
+    halide_debug_assert(user_context, false && "unimplemented");
+    return halide_error_code_unimplemented;
+}
+
+WEAK int halide_xrt_initialize_kernels(void *user_context, void **state_ptr, const char *kernel_name) {
+    int ret;
+    char xclbin_name[512];
+    char open_name[512];
+    xuid_t uuid;
+    xrtKernelHandle kernel_handle;
+    XrtKernelState *state;
+
+    debug(user_context)
+        << "XRT: halide_xrt_initialize_kernels (user_context: " << user_context
+        << ", state_ptr: " << state_ptr
+        << ", kernel_name: " << kernel_name << ")\n";
+
+    XRTContext context(user_context);
+    if (context.error_code)
+        return context.error_code;
+
+    snprintf(xclbin_name, sizeof(xclbin_name), "%s.xclbin", kernel_name);
```

Comment:
Here and elsewhere, should check the result code for an error -- IIRC, a negative number returned if buffer too small.